arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/NonEmptySet.kt

```diff
@@ -0,0 +1,34 @@
+package arrow.core
+
+public class NonEmptySet<out T>(
+  private val first: T,
+  private val rest: Set<T>,
+) : AbstractSet<T>() {
+
+  public val elements: Set<T> by lazy { setOf(first) + rest }
+
+  public operator fun plus(other: NonEmptySet<@UnsafeVariance T>): NonEmptySet<T> =
+    NonEmptySet(this.first, rest + other.elements)
+
+  public operator fun plus(s: Set<@UnsafeVariance T>): NonEmptySet<T> =
+    NonEmptySet(first, rest + s)
+
+  public operator fun plus(s: @UnsafeVariance T): NonEmptySet<T> =
+    NonEmptySet(first, rest + s)
+
+  override val size: Int
+    get() = elements.size
+
+  override fun iterator(): Iterator<T> = elements.iterator()
+
+  override fun toString(): String = "NonEmptySet(${elements.joinToString()})"
+}
+
+public fun <T> nonEmptySetOf(fist: T, vararg rest: T): NonEmptySet<Any?> =
+  NonEmptySet(fist, rest.toSet())
+
+public fun <T> Collection<T>.toNonEmptySetOrNull(): NonEmptySet<T>? =
+  firstOrNull()?.let { NonEmptySet(it, this.drop(1).toSet()) }
```

Comment:
Does this require the source collection to have a stable iteration order?

Reply:
Not sure. I mean, the order is not relevant for a set (unless is a sorted set), right?
I think your concern is that this function may not return the exact same output every time, right?

Reply:
I think this code relies on an assumption that the element dropped by `drop(1)` will be the same element that was returned by `firstOrNull`. That would require the iterator to return its elements in the same order each time. Although that's probably true for the common standard library collections, I don't think collection iterators are required to guarantee it.

Reply:
I get your point, maybe this is better then:
```kotlin
firstOrNull()?.let { NonEmptySet(it, minus(it).toSet()) }
```